### PR TITLE
#1613 - feat(visualization): add robust direct route navigation

### LINKED ISSUE #1487: Support following direct steps
### Please describe the feature that you want to propose

maybe by double-clicking on a direct step, it would be nice to select an dput focus to the target direct element.

For consumer, take care that there can have many steps and we would need to let user choose.

This can be done in the same file initially.

Another iteration searching in a whole workspace.

### LINKED ISSUE #1613: Multi-route navigation
### Please describe the feature that you want to propose

In camel, it's common to link one (to) direct endpoint to its counterpart (from). i.e.

from(rest:topLevel)
// some processing
to(direct:lowerLevel)

And another route would be defined in the same context, as follows:

from(direct:lowerLevel)
//some steps

It would be nice if users could navigate between the to(direct:lowerLevel) and from(direct:lowerLevel) by clicking on some sort of "link" icon (attached) and then he would be redirected to the corresponding lowerLevel route. And from lowerLower route it would be possible to be redirected with ease to the parent route..

image

### PR DESCRIPTION
**Pull request by lhein**
Added an icon button to direct endpoints on the canvas which acts as a "Go To" functionality. When clicking the button on a To node the selection changes to the referenced direct from endpoint. If the route containing this element is not visible, it will be made visible. There is also the reverse functionality where you can jump from a direct from endpoint to the caller. If there are multiple callers then a popover is shown which lets the user select the right route. 

- Add direct endpoint normalization for both URI notations:
  - `uri: direct` + `parameters.name`
  - `uri: direct:<name>` (including query params)
- Resolve direct links from both route and REST callers
- Add reverse navigation from `from(direct:...)` back to caller(s)
- Show a route chooser popover when multiple callers exist
- Keep single-caller behavior as one-click navigation
- Hide popover close icon and improve chooser UX (button-style entries, plain list)
- Fix stale navigation icon updates after source changes
- Extend unit tests for URI parsing and direct route navigation flows

Fixes #1613
Fixes #1487

### CODE REVIEW

**Comment by lhein** [comment]
https://github.com/user-attachments/assets/007efdca-4922-4d96-9dd8-acea66ae3c9c

File: packages/ui/src/utils/camel-uri-helper.ts [hunk 1]
```diff
@@ -36,19 +37,54 @@ export class CamelUriHelper {
...
     switch (camelElementLookup.componentName) {
       case 'direct':
-        return getValue(value, 'parameters.name');
+        return this.getDirectEndpointName(value);
     }
 
     return undefined;
   }
 
+  static getDirectEndpointName<T>(value: T | undefined | null): string | undefined {
```
Comment (lordrip) [inline packages/ui/src/utils/camel-uri-helper.ts]: Is this necessary? We're not using the generic, since the value could be string or an object 🤔

File: packages/ui/src/components/Visualization/Custom/Node/direct-route-navigation.ts
Comment (lordrip) [inline packages/ui/src/components/Visualization/Custom/Node/direct-route-navigation.ts:1]: Ideally, this file should be a class and not a set of functions. Also, we should move it to the camel folder, since this is only necessary there

File: packages/ui/src/components/Visualization/Custom/Node/CustomNodeContent.tsx [hunk 2]
```diff
@@ -46,6 +75,44 @@ export const CustomNodeContent: FunctionComponent<CustomNodeContentProps> = ({
...
         </FloatingCircle>
       )}
+      {(onNavigateToRoute || hasNavigateOptions) && (
+        <FloatingCircle className="step-icon step-icon__navigation">
+          {hasNavigateOptions ? (
+            <Popover
+              triggerAction="click"
+              isVisible={isNavigatePopoverVisible}
+              shouldOpen={() => setIsNavigatePopoverVisible(true)}
+              shouldClose={() => setIsNavigatePopoverVisible(false)}
```
Comment (lordrip) [inline packages/ui/src/components/Visualization/Custom/Node/CustomNodeContent.tsx]:
```suggestion
              shouldOpen={() => { setIsNavigatePopoverVisible(true); }}
              shouldClose={() => { setIsNavigatePopoverVisible(false); }}
```

File: packages/ui/src/components/Visualization/Custom/Node/CustomNodeContent.tsx [hunk 3]
```diff
@@ -22,7 +26,32 @@ export const CustomNodeContent: FunctionComponent<CustomNodeContentProps> = ({
...
+  onNavigateToRoute,
+  navigateToRouteOptions,
+  onNavigateToRouteSelect,
 }) => {
+  const [isNavigatePopoverVisible, setIsNavigatePopoverVisible] = useState(false);
+  const hasNavigateOptions =
+    Array.isArray(navigateToRouteOptions) && navigateToRouteOptions.length > 1 && !!onNavigateToRouteSelect;
+
+  const navigateButton = (
+    <button
```
Comment (lordrip) [inline packages/ui/src/components/Visualization/Custom/Node/CustomNodeContent.tsx]: shouldn't we use the patternfly button instead?

File: packages/ui/src/components/registers/RegisterComponents.tsx [hunk 4]
```diff
@@ -20,6 +21,15 @@ export const RegisterComponents: FunctionComponent<PropsWithChildren> = ({ child
...
       component: lazy(() => import('../ComponentMode/ComponentMode')),
     },
+    {
+      anchor: Anchors.CanvasNodeBottomRight,
+      activationFn: directRouteNavigationActivationFn,
+      component: lazy(() =>
+        import('../Visualization/Custom/Node/DirectRouteNavigationAnchor').then((module) => ({
+          default: module.DirectRouteNavigationAnchor,
+        })),
+      ),
```
Comment (PVinaches) [inline packages/ui/src/components/registers/RegisterComponents.tsx:31]:
```
...
    component: lazy(() => import('../Visualization/Custom/Node/DirectRouteNavigationAnchor')),
},
```

File: packages/ui/src/components/registers/direct-route-navigation.activationfn.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import { ProcessorDefinition } from '@kaoto/camel-catalog/types';
+
+import { IVisualizationNode } from '../../models';
+import { DirectRouteNavigationService } from '../../models/camel/direct-route-navigation.service';
+import { CamelRouteVisualEntityData } from '../../models/visualization/flows/support/camel-component-types';
+
+export const directRouteNavigationActivationFn = (vizNode: IVisualizationNode): boolean => {
```
Comment (PVinaches) [inline packages/ui/src/components/registers/direct-route-navigation.activationfn.ts:7]:
```
const directRouteNavigationActivationFn = (vizNode: IVisualizationNode): boolean => {
...
```

File: packages/ui/src/components/registers/direct-route-navigation.activationfn.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+  const isDirectToNode = nodeData.processorName === 'to' && nodeData.componentName === 'direct';
+  const isDirectFromNode =
+    nodeData.processorName === ('from' as keyof ProcessorDefinition) && nodeData.componentName === 'direct';
+
+  if (!isDirectToNode && !isDirectFromNode) {
+    return false;
+  }
+
+  return DirectRouteNavigationService.getDirectEndpointNameFromDefinition(vizNode.getNodeDefinition()) !== undefined;
+};
```
Comment (PVinaches) [inline packages/ui/src/components/registers/direct-route-navigation.activationfn.ts:18]:
```
...
export default directRouteNavigationActivationFn;
```

File: packages/ui/src/components/registers/direct-route-navigation.activationfn.test.ts [hunk 7]
```diff
@@ -0,0 +1,131 @@
+import { ProcessorDefinition } from '@kaoto/camel-catalog/types';
+
+import { IVisualizationNode } from '../../models';
+import { DirectRouteNavigationService } from '../../models/camel/direct-route-navigation.service';
```
Comment (PVinaches) [inline packages/ui/src/components/registers/direct-route-navigation.activationfn.test.ts:4]:
```
import directRouteNavigationActivationFn from './direct-route-navigation.activationfn';
...
```

File: packages/ui/src/components/registers/direct-route-navigation.activationfn.test.ts [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+    jest.spyOn(DirectRouteNavigationService, 'getDirectEndpointNameFromDefinition').mockReturnValue('myEndpoint');
+
+    const result = directRouteNavigationActivationFn(vizNode);
+
+    expect(result).toBe(true);
+    expect(DirectRouteNavigationService.getDirectEndpointNameFromDefinition).toHaveBeenCalledWith(nodeDefinition);
+  });
+
+  it('should return true for direct "from" node with valid endpoint name', () => {
```
Comment (PVinaches) [inline packages/ui/src/components/registers/direct-route-navigation.activationfn.test.ts:63]: Redundant as it covers the same code path that the previous test

File: packages/ui/src/components/registers/direct-route-navigation.activationfn.test.ts [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+    jest.spyOn(DirectRouteNavigationService, 'getDirectEndpointNameFromDefinition').mockReturnValue(undefined);
+
+    const result = directRouteNavigationActivationFn(vizNode);
+
+    expect(result).toBe(false);
+    expect(DirectRouteNavigationService.getDirectEndpointNameFromDefinition).toHaveBeenCalledWith(nodeDefinition);
+  });
+
+  it('should return false for direct "from" node without endpoint name', () => {
```
Comment (PVinaches) [inline packages/ui/src/components/registers/direct-route-navigation.activationfn.test.ts:99]: Redundant from the same reason (already covered with previous to test)

File: packages/ui/src/components/registers/direct-route-navigation.activationfn.test.ts [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+    jest.spyOn(DirectRouteNavigationService, 'getDirectEndpointNameFromDefinition').mockReturnValue(undefined);
+
+    const result = directRouteNavigationActivationFn(vizNode);
+
+    expect(result).toBe(false);
+    expect(DirectRouteNavigationService.getDirectEndpointNameFromDefinition).toHaveBeenCalledWith(nodeDefinition);
+  });
+
+  it('should handle edge case with empty componentName', () => {
```
Comment (PVinaches) [inline packages/ui/src/components/registers/direct-route-navigation.activationfn.test.ts:117]: This is basically the same than the first test with a 'http'. I think that the first test can be refactored to cover this too.

File: packages/ui/src/components/registers/direct-route-navigation.activationfn.test.ts [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+  });
+
+  it('should return false if vizNode data is not a direct component', () => {
+    const vizNode = {
+      data: {
+        processorName: 'to',
+        componentName: 'http',
+      } as CamelRouteVisualEntityData,
+      getNodeDefinition: jest.fn().mockReturnValue({}),
+    } as unknown as IVisualizationNode;
```
Comment (PVinaches) [inline packages/ui/src/components/registers/direct-route-navigation.activationfn.test.ts:22]: suggestion to avoid duplication:

```
const makeVizNode = (
  processorName: string,
  componentName: string,
  nodeDefinition: object = {},
): IVisualizationNode =>
  ({
    data: { processorName, componentName } as CamelRouteVisualEntityData,
    getNodeDefinition: jest.fn().mockReturnValue(nodeDefinition),
  }) as unknown as IVisualizationNode;
```

File: packages/ui/src/components/registers/direct-route-navigation.activationfn.ts [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+
+import { IVisualizationNode } from '../../models';
+import { DirectRouteNavigationService } from '../../models/camel/direct-route-navigation.service';
+import { CamelRouteVisualEntityData } from '../../models/visualization/flows/support/camel-component-types';
+
+export const directRouteNavigationActivationFn = (vizNode: IVisualizationNode): boolean => {
+  const nodeData = vizNode.data as CamelRouteVisualEntityData;
+  const isDirectToNode = nodeData.processorName === 'to' && nodeData.componentName === 'direct';
+  const isDirectFromNode =
+    nodeData.processorName === ('from' as keyof ProcessorDefinition) && nodeData.componentName === 'direct';
```
Comment (PVinaches) [inline packages/ui/src/components/registers/direct-route-navigation.activationfn.ts:11]: Why the 'from' needs casting? It should work fine without as in the case of `isDirectToNode` just before it. Can't we update the ProcessorDefinition to include 'from' instead?

File: packages/ui/src/components/Visualization/Custom/Node/DirectRouteNavigationAnchor.tsx [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+import { CamelRouteVisualEntityData } from '../../../../models/visualization/flows/support/camel-component-types';
+import { SourceCodeContext } from '../../../../providers/source-code.provider';
+import { VisibleFlowsContext } from '../../../../providers/visible-flows.provider';
+import { FloatingCircle } from '../FloatingCircle/FloatingCircle';
+
+interface DirectRouteNavigationAnchorProps {
+  vizNode?: IVisualizationNode;
+}
+
+export const DirectRouteNavigationAnchor: FunctionComponent<DirectRouteNavigationAnchorProps> = ({ vizNode }) => {
```
Comment (PVinaches) [inline packages/ui/src/components/Visualization/Custom/Node/DirectRouteNavigationAnchor.tsx:19]:
```
const DirectRouteNavigationAnchor: FunctionComponent<DirectRouteNavigationAnchorProps> = ({ vizNode }) => {
...
```

File: packages/ui/src/components/Visualization/Custom/Node/DirectRouteNavigationAnchor.tsx [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+          }
+        >
+          {navigateButton}
+        </Popover>
+      ) : (
+        navigateButton
+      )}
+    </FloatingCircle>
+  );
+};
```
Comment (PVinaches) [inline packages/ui/src/components/Visualization/Custom/Node/DirectRouteNavigationAnchor.tsx:255]:
```
...
export default DirectRouteNavigationAnchor;
```

File: packages/ui/src/components/Visualization/Custom/Node/DirectRouteNavigationAnchor.tsx [hunk 15]
```diff
@@ -0,0 +1,255 @@
...
+        singleTargetId: undefined,
+        targetOptions: [] as string[],
+        targetNodeIds: {} as Record<string, string>,
+      };
+    }
+    // Recompute direct navigation targets when source changes without recreating entities context.
+    void sourceCode;
+
+    const nodeData = vizNode.data as CamelRouteVisualEntityData;
+    const isDirectToNode = nodeData.processorName === 'to' && nodeData.componentName === 'direct';
```
Comment (PVinaches) [inline packages/ui/src/components/Visualization/Custom/Node/DirectRouteNavigationAnchor.tsx:40]: `isDirectToNode` and `isDirectFromNode` are duplicated in `direct-route-navigation.activationfn.ts`

File: packages/ui/src/components/Visualization/Custom/Node/DirectRouteNavigationAnchor.tsx [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+interface DirectRouteNavigationAnchorProps {
+  vizNode?: IVisualizationNode;
+}
+
+export const DirectRouteNavigationAnchor: FunctionComponent<DirectRouteNavigationAnchorProps> = ({ vizNode }) => {
+  const [isPopoverVisible, setIsPopoverVisible] = useState(false);
+  const element = useContext(ElementContext);
+  const entitiesContext = useEntityContext();
+  const visibleFlowsContext = useContext(VisibleFlowsContext);
+  const sourceCode = useContext(SourceCodeContext);
```
Comment (PVinaches) [inline packages/ui/src/components/Visualization/Custom/Node/DirectRouteNavigationAnchor.tsx:24]: This is a strange workaround that maybe can be avoided by updating updateEntitiesFromCamelResource in entities.provided.tsx instead. It doesn't need to check sourceCode to force useMemo to recompute if entitiesContext are replaced instead of mutating in place. I think it is worth looking at this.

File: packages/ui/src/components/Visualization/Custom/Node/DirectRouteNavigationAnchor.tsx [hunk 17]
```diff
@@ -0,0 +1,255 @@
...
+
+  const directNavigation = useMemo(() => {
+    if (!vizNode || !entitiesContext) {
+      return {
+        directEndpointName: undefined,
```
Comment (PVinaches) [inline packages/ui/src/components/Visualization/Custom/Node/DirectRouteNavigationAnchor.tsx:29]: I think it is cleaner to avoid the casting like this in every return with an interface:

```
interface DirectNavigation {
  directEndpointName: string | undefined;
  targetProcessorName: 'from' | 'to' | undefined;
  singleTargetId: string | undefined;
  targetOptions: string[];
  targetNodeIds: Record<string, string>;
}
```

File: packages/ui/src/components/Visualization/Custom/Node/DirectRouteNavigationAnchor.tsx (hunk 13 above)
Comment (PVinaches) [inline packages/ui/src/components/Visualization/Custom/Node/DirectRouteNavigationAnchor.tsx:19]: This file is a little complex to follow. What do you think of some refactoring to hooks? For example refactoring into useDirectNavigation and useNavigateToRoute. It will also simplify and isolate the testing (easier debugging later too)

**Review by PVinaches — COMMENTED** [review]
I reviewed and tested this PR until models and utils (I can go back to them later). In my end it didn't appear the popover nor worked from one direct to multiple options (but yes the other way around). I sent screenshots to @lordrip about this. I think it still needs some work on it (refactoring, test simplification among other). I hope my comments help with this :)

**Comment by lhein** [comment]
closing the PR until we decide how to proceed
